### ISSUE TITLE
Data adapter: Broadcasting causes issue with sampling

### ISSUE THREAD

**Issue by maltelueken** [bayesflow-org/bayesflow#258]
When using a `Adapter` with a `Broadcast` transform where `to="inference_variables"`, training runs normally, but when I try sampling it complains that it cannot find the key `"inference_variables"` which I did not include from the `conditions` dictionary following the example notebooks.
My guess is the `Broadcast` transform also needs a `strict` argument in the `forward` method for sampling as in the `FilterTransform` class?

**Comment by paul-buerkner** [comment]
An you show the simulator and adapter you used? I will then check what we
can do about it.

Malte Lüken ***@***.***> schrieb am Fr., 22. Nov. 2024, 14:55:

**Comment by paul-buerkner** [comment]
thanks!

Malte Lüken ***@***.***> schrieb am Fr., 22. Nov. 2024, 15:41:

**Comment by paul-buerkner** [comment]
I have checked and I don't think a `strict` argument would fix it. The issue is that without the inference variables (inferred during `approximator.sample`), we cannot broadcast properly, which we need for the forward pass. In other words, the broadcast `to` argument should not receive inference variables. I will add this to the documentation. For your example, what would be the easiest way for you to broadcast as you intented without using `to = "inference_variables`?

**Comment by maltelueken** [comment]
Thanks for looking into this! I could also modify my prior function (which requires broadcasting) to achieve the same result, but I thought using the broadcasting transform would be more elegegant. Perhaps it would be good to add an explicit exception when people use `to="inference_variables"`? The error message that appears is not super informative on what the actual issue is I think.

**Comment by paul-buerkner** [comment]
Yeah, something like this could make sense. Even better, we would need to throw an error whenever a variable is used for broadcasting that includes an inferred variable, which will not be present during inference time at the time broadcasting is called. It may take a while until we can implement this smart checking as we have to wait for other features but I think it would be very nice to have indeed.

**Comment by paul-buerkner** [comment]
I have added a note in the doc of broadcast to warn about this case. I don't think we can easily inspect the adapter to automatically prohibit this case in the first place, so I consider the doc solution to be fine for now. If users continue to run into this issue, we can reconsider a more sophisticated solution.
